seis/ms/header.go

```diff
@@ -0,0 +1,359 @@
+	for len(s) < 5 {
+		s += " "
+	}
+	copy(h.StationIdentifier[:], s)
+}
+func (h RecordHeader) Location() string {
+	return strings.TrimSpace(string(h.LocationIdentifier[:]))
+}
+func (h *RecordHeader) SetLocation(s string) {
+	for len(s) < 2 {
+		s += " "
+	}
+	copy(h.LocationIdentifier[:], s)
+}
+func (h RecordHeader) Channel() string {
+	return strings.TrimSpace(string(h.ChannelIdentifier[:]))
+}
+func (h *RecordHeader) SetChannel(s string) {
+	for len(s) < 3 {
+		s += " "
+	}
+	copy(h.ChannelIdentifier[:], s)
+}
+func (h RecordHeader) Network() string {
+	return strings.TrimSpace(string(h.NetworkIdentifier[:]))
+}
+func (h *RecordHeader) SetNetwork(s string) {
+	for len(s) < 2 {
+		s += " "
+	}
+	copy(h.NetworkIdentifier[:], s)
+}
+func (h RecordHeader) SrcName(quality bool) string {
+	if quality {
+		return strings.Join([]string{h.Network(), h.Station(), h.Location(), h.Channel(), string(h.DataQualityIndicator)}, "_")
+	}
+	return strings.Join([]string{h.Network(), h.Station(), h.Location(), h.Channel()}, "_")
+}
+
+func (h *RecordHeader) SetStartTime(t time.Time) {
+	h.RecordStartTime = NewBTime(t)
+}
+
+func (h *RecordHeader) SetCorrection(correction time.Duration, applied bool) {
+	switch applied {
+	case true:
+		h.ActivityFlags = setBit(h.ActivityFlags, 1)
+	default:
+		h.ActivityFlags = clearBit(h.ActivityFlags, 1)
+	}
+	h.TimeCorrection = int32(correction / (100 * time.Microsecond))
+}
+
+func (h RecordHeader) Correction() time.Duration {
+	return 100 * time.Microsecond * time.Duration(h.TimeCorrection)
+}
+
+func (h RecordHeader) StartTime() time.Time {
+
+	start := h.RecordStartTime.Time()
+
+	// Check for a header time correction
+	if !isBitSet(h.ActivityFlags, 1) { // bit 1 indicates whether time correction has been applied
+		start = start.Add(100 * time.Microsecond * time.Duration(h.TimeCorrection))
+	}
+
+	return start
+}
+
+// SampleRate returns the decoded header sampling rate in samples per second.
+func (h RecordHeader) SampleRate() float64 {
+	return sampleRate(int(h.SampleRateFactor), int(h.SampleRateMultiplier))
+}
+
+// SampleCount returns the number of samples in the record, independent of whether they are decoded or not.
+func (h RecordHeader) SampleCount() int {
+	return int(h.NumberOfSamples)
+}
+
+// SamplePeriod converts the sample rate into a time interval, or zero.
+func (h RecordHeader) SamplePeriod() time.Duration {
+	return samplePeriod(int(h.SampleRateFactor), int(h.SampleRateMultiplier))
+}
+
+// IsValid performs a simple consistency check of the RecordHeader contents.
+func (h RecordHeader) IsValid() bool {
+	for _, b := range h.SequenceNumber {
+		if !((b >= '0' && b <= '9') || (b == ' ') || (b == 0)) {
+			return false
+		}
+	}
+
+	if q := h.DataQualityIndicator; !(q == 'D' || q == 'R' || q == 'M' || q == 'Q') {
+		return false
+	}
+
+	if b := h.ReservedByte; !(b == ' ' || b == 0) {
+		return false
+	}
+
+	if s := h.RecordStartTime; !(s.Hour <= 23 && s.Minute <= 59 && s.Second <= 60) {
```

Comment:
`s.Second <= 59` ?

Reply:
when there's a leap second the number of seconds can be 60 (not sure what would happen if there were more than one though!)

From the SEED manual.

```
Field type Number of bits Field description
UWORD 16 Year (e.g., 1987)
UWORD 16 Day of Year (Jan 1 is 1)
UBYTE 8 Hours of day (0—23)
UBYTE 8 Minutes of day (0—59)
UBYTE 8 Seconds of day (0—59, 60 for leap seconds)
UBYTE 8 Unused for data (required for alignment)
UWORD 16 .0001 seconds (0—9999)
```